jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,407 @@
+    init_x = env.observation_spec().generate_value()
+    # select only obs for a single agent.
+    init_x = jax.tree_util.tree_map(lambda x: x[0], init_x)
+    init_x = jax.tree_util.tree_map(lambda x: x[None, ...], init_x)
+    params = network.init(rng_p, init_x)
+    # initialise optimiser stats.
+    opt_state = optim.init(params)
+
+    vmapped_network_apply_fn = jax.vmap(
+        network.apply, in_axes=(None, 1), out_axes=(1, 1)
+    )
+
+    learn = get_learner_fn(  # get batched iterated update.
+        env, vmapped_network_apply_fn, optim.update, config
+    )
+    learn = jax.pmap(learn, axis_name="device")  # replicate over multiple cores.
+
+    # BROADCAST PARAMS AND OPTIMISER STATE
+    broadcast = lambda x: jnp.broadcast_to(
+        x, (cores_count, config["UPDATE_BATCH_SIZE"]) + x.shape
+    )
+    params = jax.tree_map(broadcast, params)  # broadcast to cores and batch.
+    opt_state = jax.tree_map(broadcast, opt_state)  # broadcast to cores and batch
+
+    # INITIALISE ENVIRONMENT
+    rng, *env_rngs = jax.random.split(
+        rng, cores_count * config["UPDATE_BATCH_SIZE"] * config["NUM_ENVS"] + 1
+    )
+    env_states, timesteps = jax.vmap(env.reset, in_axes=(0))(
+        jnp.stack(env_rngs),
+    )
+    rng, *step_rngs = jax.random.split(
+        rng, cores_count * config["UPDATE_BATCH_SIZE"] + 1
+    )
+
+    # RESHAPE ENVIRONMENT STATES AND TIMESTEPS
+    reshape_step_rngs = lambda x: x.reshape(
+        (cores_count, config["UPDATE_BATCH_SIZE"]) + x.shape[1:]
+    )
+    step_rngs = reshape_step_rngs(jnp.stack(step_rngs))  # add dimension to pmap over.
+
+    reshape_states = lambda x: x.reshape(
+        (cores_count, config["UPDATE_BATCH_SIZE"], config["NUM_ENVS"]) + x.shape[1:]
+    )
+    env_states = jax.tree_util.tree_map(
+        reshape_states, env_states
+    )  # add dimension to pmap over.
+    timesteps = jax.tree_util.tree_map(
+        reshape_states, timesteps
+    )  # add dimension to pmap over.
+
+    # RUN EXPERIMENT
+    with TimeIt(tag="COMPILATION"):
+        learn(params, opt_state, step_rngs, env_states, timesteps)
+
+    with TimeIt(tag="EXECUTION", environment_steps=total_timesteps):
+        output = learn(params, opt_state, step_rngs, env_states, timesteps)
+
+    return output
+
+
+if __name__ == "__main__":
+    config = {
+        "LR": 2.5e-4,
+        "UPDATE_BATCH_SIZE": 4,
+        "ROLLOUT_LENGTH": 128,
+        "NUM_UPDATES": 10,
+        "NUM_ENVS": 32,
+        "PPO_EPOCHS": 4,
+        "NUM_MINIBATCHES": 8,
+        "GAMMA": 0.99,
+        "GAE_LAMBDA": 0.95,
+        "CLIP_EPS": 0.2,
+        "ENT_COEF": 0.01,
+        "VF_COEF": 0.5,
+        "MAX_GRAD_NORM": 0.5,
+        "ACTIVATION": "relu",
+        "ENV_NAME": "RobotWarehouse-v0",
+        "SEED": 42,
+    }
+
+    env = jumanji.make(config["ENV_NAME"])
+    env = RwareMultiAgentWrapper(env)
+    env = AutoResetWrapper(env)
+    env = LogWrapper(env)
+
+    output = run_experiment(env, config)
```

Comment:
Not using output here, so either just do `run_experiment(env, config)` or do some sort of final print out using the output. Should probably have a final print anyways, like `print("IPPO experiment completed")`